added two tests that replace test_atomic_reprepare()

mishinma opened the pull request:
@yeganer @unoebauer @wkerzendorf 
Hi, this is my first pull request! I've implemented the idea from #473 . 
- Firstly, I removed the  `pytest.mark.skipif` marker and moved all the verification to the `atomic_data_fname` fixture in the tardis/conftest.py. This way we don't have to check it in each test that requests an atomic database. I also included two other fixtures: `lines_dataset` that returns  lines data from hdf5 file and `atom_data_from_dataset` that requests the `atomic_data_fname` fixture and returns an AtomData object.
- Then I replaced `test_atomic_reprepare` with two tests: `test_prepare_atom_data_set_lines` and `test_prepare_atom_data_set_lines_w_max_ion_number`. I decided to go with to tests because I didn't want to overload them with logic - they are unit tests after all. So the second one does basically the same thing as the first but also checks max_ion_number. As suggested in #473 the number of lines is calculated from the provided atomic data file and then compared with the value from `prepare_atom_data` method. 

Look forward to a review

yeganer commented on tardis/conftest.py:
```diff
@@ -145,20 +145,43 @@ def pytest_report_header(config):
+
+
+@pytest.fixture(scope="session")
+def lines_dataset(atomic_data_fname, request):
```
I think this fixture belongs to test_atomic.py

yeganer commented on tardis/conftest.py:
```diff
@@ -145,20 +145,43 @@ def pytest_report_header(config):
 from tardis.atomic import AtomData
 
 @pytest.fixture
+def atom_data_from_dataset(atomic_data_fname):
```
This fixture should be moved to test_atomic.py, too. Accessing atom_data without calling atom_data.prepare_atom_data is dangerous.

yeganer commented on tardis/tests/test_atomic.py:
```diff
@@ -38,22 +38,40 @@ def test_atomic_symbol():
-    atom_data.prepare_atom_data([20])
-    assert len(atom_data.lines) > 0
+
+param_atom_num = [[14], [20], [14, 20]]
```
I wonder if this is needed. Something like
```
@pytest.mark.parametrize("selected_atomic_numbers", [[14], [20], [14, 20]])
```
should to the trick, right? (Haven't tested that but the goal should be something like that.)

mishinma replied:
Sure, but I wanted to use ids to make test names more informative.  
In the latest version of pytest ids can be either a list of string names or a function that returns names, but the version used by Travis seems only to support list of strings. So without using a global variable it could be like this: 
```
@pytest.mark.parametrize("selected_atomic_numbers", [[14], [20], [14, 20]],
  pytest.config.getvalue('atomic-dataset')))
                       ids = ["atom_num: {}".format(_) for _ in [[14], [20], [14, 20]]])
```
Without ids at all pytest doesn't show atomic numbers, only the name of the list with index:
`tardis/tests/test_atomic.py::test_prepare_atom_data_set_lines[selected_atomic_numbers0] PASSED`\
Well, that was my reasoning, but I can remove ids altogether.

yeganer commented:
@mishinma Thank you for the contribution. I really like your documentation of the code.

I've added some comments to the code. But it looks already really good.

mishinma commented:
@yeganer I moved both fixtures to test_atomic.py. And
- used yield fixture mechanism to handle hdf5 in the lines_dataset fixture
- changed the name of the second one to atom_data_f_dataset_not_prepared

mishinma commented:
After giving it more thought I removed the `atom_data_f_dataset_not_prepared` fixture completely - it only adds ambiguity.

yeganer commented:
That looks really good.

@wkerzendorf @unoebauer Can you please have a look and merge if you have no further comments?

unoebauer commented:
Looks very good! I am however confused about our decrease in coverage, given that we have actually added and improved tests with this PR.

In any case, having looked over the source code changes, I'm fine with merging this one. @wkerzendorf, what do you think?

yeganer commented:
The decrease in coverage comes from parameterized tests not being
recognized.

Ulrich Noebauer <notifications@github.com> schrieb am Di., 9. Feb. 2016
10:33:

> Looks very good! I am however confused about our decrease in coverage,
> given that we have actually added and improved tests with this PR.
>
> In any case, having looked over the source code changes, I'm fine with
> merging this one. @wkerzendorf <https://github.com/wkerzendorf>, what do
> you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tardis-sn/tardis/pull/476#issuecomment-181780626>.
>

wkerzendorf commented:
@embray - did you setup coveralls for astropy? Does astropy also have a problem with parameterized tests not appearing in the coverage?

unoebauer commented:
@wkerzendorf - what to do with this one (extends to #473)?

unoebauer commented:
Closing this one - @wkerzendorf reopen if you think this is still relevant.
